### ISSUE TITLE
Training with max_d 256

### ISSUE THREAD

**Issue by shubhamag** [YoYo000/MVSNet#17]
Hi,

I was trying to compare the performance of the net with different number of depth samples. My expectation (and as per what the paper suggests) would be that higher number of depth samples would lead to more accurate depth predictions.
I ran two experiments on DTU, one with max_d 128, interval_scale set to 1.6 , and second with max_d 256 and interval_scale set to 0.8. I observe that the first model is performing much better (% < 1mm close to 90% )  than the second model with higher depth samples (%<1mm around 80% ) .

Is there anything else in the code to be changed when training/testing with different max_d besides interval_scale ? I am using the provided DTU cam files.

**Comment by YoYo000** [comment]
That is an interesting discovery. I also notice that setting D=128 produces satisfying training result. but I did not quantitatively evaluate their differences.

I think one possibility is H, W and D should be with the same level of resolutions. During training the input images are downsized by 2, so maybe it is better to also downsize the depth resolution from 2mm to 4mm.

**Comment by tejaskhot** [comment]
@YoYo000 From the README, the maximum depth is computed as:
`DEPTH_MAX = DEPTH_MIN + (interval_scale * DEPTH_INTERVAL) * (max_d - 1)`
For this equation to hold with 256 depth samples, the other values need to be:
```
DEPTH_MAX = 935
DEPTH_MIN = 425
interval_scale = 0.8
DEPTH_INTERVAL = 2.5
max_d = 256
```

Can you please clarify the exact train and test setup on DTU for getting optimal results? 
From the paper, it seemed like the train setting was: 
`max_d=256`, `interval_scale=0.8`, `max_w=640`, `max_h=512`
and for test:
`max_d=256`, `interval_scale=0.8`, `max_w=1600`, `max_h=1184`

But from the experiments, it seems like this train setup (above) is worse than:
`max_d=128`, `interval_scale=1.6`, `max_w=640`, `max_h=512`

**Comment by YoYo000** [comment]
@tejaskhot In the paper I used the training/testing setup as you mentioned, but did not benchmark the training with max_d=128, interval_scale=1.6. For testing, I have done an ablation study proving D=256 is the optimal setup ([supp](http://www.cse.ust.hk/~yyaoag/papers/yao2018mvsnet_supp.pdf))

It is good to see that using small D leads to better training results, which means we can train faster with less GPU memory. I am also recently fine-tuning the network (speed, memory and structure fine-tuning). Could you share your statistics on depth sample number and depth interval?

Thanks

**Comment by tejaskhot** [comment]
Thanks for the quick response. Can you please describe what you are referring to as the optimal setup? I think from Sec 3.1 of the supplemental, you mean that you train with `max_d=128, interval_scale=1.6, max_w=640, max_h=512` but then vary the `max_d` and find `256` to be the best. Is that right?

I was under the impression that you used `max_d=256` during training time too. I am not changing the depth sample number and it is kept as is at `2.5`. I only change `max_d` and `interval_scale` as mentioned in the previous comment and all other settings are kept the same.

**Comment by YoYo000** [comment]
In the paper, all trainings are performed with d=256 and interval_scale=0.8
(2mm)

Tejas Khot <notifications@github.com> 于 2018年12月20日周四 上午11:49写道：

**Comment by YoYo000** [comment]
> Thanks for the quick response. Can you please describe what you are referring to as the optimal setup? I think from Sec 3.1 of the supplemental, you mean that you train with `max_d=128, interval_scale=1.6, max_w=640, max_h=512` but then vary the `max_d` and find `256` to be the best. Is that right?
> 
> I was under the impression that you used `max_d=256` during training time too. I am not changing the depth sample number and it is kept as is at `2.5`. I only change `max_d` and `interval_scale` as mentioned in the previous comment and all other settings are kept the same.

Sec 3.1. studies the influence of D on testing rather than training. All experiments in the paper and supp shares one single model trained with d=256 and interval_scale=0.8.

**Comment by tejaskhot** [comment]
I think I am a bit confused here. Isn't the setup described by @shubhamag in the original issue comment same as the optimal one? I guess the issue is that we are not able to get the corresponding improvement like pointed out in the original post. Please correct me if wrong.

**Comment by YoYo000** [comment]
In the paper the **training** is performed with 'max_d=256, interval_scale=0.8, max_w=640, max_h=512'. and the ablation study in supp Sec. 3.1 uses this model without any change.

Sec 3.1. is to study the influence of 'max_d' and 'interval_scale' on **testing** (with fixed 'DEPTH_MAX=935, DEPTH_MIN=425, max_w=1600, max_h=1184'), and it shows 'max_d=256, interval_scale=0.8, max_w=1600, max_h=1184' is the optimal setup (given the model trained with 'max_d=256, interval_scale=0.8, max_w=640, max_h=512')

But what I cannot sure is whether training the model with 'max_d=128, interval_scale=1.6, max_w=640, max_h=512' can do better as I haven't done any quantitatively study on **training** (although the training script sets 'max_d=128', it is just training on GPU with <=8 GB memory). And I think @shubhamag and you have found that it indeed improves the point cloud quality?

**Comment by YoYo000** [comment]
In the paper the **training** is performed with `max_d=256, interval_scale=0.8, max_w=640, max_h=512`. and the ablation study in supp Sec. 3.1 uses this model without any change.

Sec 3.1. is to study the influence of `max_d` and `interval_scale` on **testing** (with fixed `DEPTH_MAX=935, DEPTH_MIN=425, max_w=1600, max_h=1184`), and it shows `max_d=256, interval_scale=0.8, max_w=1600, max_h=1184` is the optimal setup (given the model trained with `max_d=256, interval_scale=0.8, max_w=640, max_h=512`)

But what I cannot sure is whether training the model with `max_d=128, interval_scale=1.6, max_w=640, max_h=512` can do better as I haven't done any quantitatively study on **training** (although the training script sets `max_d=128`, it is just training on GPU with <=8 GB memory). And I think @shubhamag and you have found that it indeed improves the point cloud quality?

**Comment by shubhamag** [comment]
Hi @YoYo000 , 
I haven't evaluated the final point cloud numbers using the DTU script yet, my observations are only from the training/validation losses of MVSNet - i.e., for the metrics L1 loss, %<1mm and %<3mm, the model with max_d = 128, interval =1.6 seems to be much better than one with max_d =256, interval = 0.8

**Comment by YoYo000** [comment]
@shubhamag @tejaskhot In the past few days I also tried training the network (a modified version to the paper) with d=128 and d=256, but what I observed is that setting d=256 is slightly better than d=128 (L1 loss is ~10% lower)

**Comment by shubhamag** [comment]
@YoYo000 was this with the DTU camera files you provide, with max_d 256, interval =0.8 vs max_d =128, interval = 1.6 ?

**Comment by nickponline** [comment]
I think I might be seeing something similar. I'm currently training on the DTU dataset:

`max_d=128`, `interval_scale=1.6`, `max_w=640`, `max_h=512`

**Comment by nickponline** [comment]
@shubhamag @tejaskhot @YoYo000 I'm looking at the results of both `max_d=128, interval_scale=1.6` and `max_d=256, interval_scale=0.8`  trained networks on the Tanks and Temples data: 

Looks like the later is generating better results though:

https://www.dropbox.com/s/5ch3hki99t53cxj/PointClouds.zip?dl=0

**Comment by shubhamag** [comment]
@nickponline Hi ! thanks for your input ... could you clarify/confirm the following ? :
1) You trained the two models on DTU ? during training do you remember which one performed better in terms of training losses ?

2) When testing on TnT, did you test both with the same settings (i.e set max_d =256,interval 0.8 for both 
 models during test ) , or test them with their corresponding train setting (i.e 128,1.6 and 256,0.8 respectively)

Thanks !

**Comment by shubhamag** [comment]
@nickponline No, I havn't been able to do that either.
Could you confirm my question regarding the comparison between training losses of the 128 and 256 max_d models when you get the chance ?
Thanks !

**Comment by kwea123** [comment]
Did you use the same `batch_size` for training both `D=256` and `D=128`? It is well known that larger batch size leads to better result, so with a given `D`, if you max the batch size on a fixed GPU memory, it is quite possible that `D=128` perform better since you can set 2x batch size than that for `D=256`. Especially for this model which consumes large memory, batch size of `2` compared to `1` can make a huge difference, so I think maybe lower D is likely to do better.
